File: algorithm_library/python/math/PascalsTriangle.py
class PascalsTriangle:
    """
    パスカルの三角形で nCk を求める
    計算量: O(N^2)
    mod が素数でなくても使用可能
    """
    def __init__(self, N, mod):
        self.mod = mod
        self.binom = [[0] * (N + 1) for _ in range(N + 1)]
        for i in range(N + 1):
            self.binom[i][0] = 1
            for j in range(1, i + 1):
                self.binom[i][j] = (self.binom[i - 1][j - 1] + self.binom[i - 1][j]) % mod

    def C(self, n, k):
        assert 0 <= k <= n
        return self.binom[n][k]

    def H(self, n, k):
        """ 重複組合せ nHk """
        assert n >= 0 and k >= 0
        return self.C(n + k - 1, k)

File: algorithm_library/cpp-from-py/math/PascalsTriangle.cpp
#include <bits/stdc++.h>
using namespace std;

struct PascalsTriangle {
    long long mod;
    vector<vector<long long>> binom;

    PascalsTriangle(int N, long long mod_) : mod(mod_), binom(N + 1, vector<long long>(N + 1, 0)) {
        for (int i = 0; i <= N; i++) {
            binom[i][0] = 1;
            for (int j = 1; j <= i; j++) binom[i][j] = (binom[i - 1][j - 1] + binom[i - 1][j]) % mod;
        }
    }

    long long C(int n, int k) const {
        assert(0 <= k && k <= n);
        return binom[n][k];
    }

    long long H(int n, int k) const {
        assert(n >= 0 && k >= 0);
        return C(n + k - 1, k);
    }
};
